pageserver/src/walredo.rs

```diff
@@ -275,17 +281,14 @@ impl PostgresRedoManager {
                 // Avoid concurrent callers hitting the same issue.
                 // We can't prevent it from happening because we want to enable parallelism.
                 {
-                    let mut guard = self.redo_process.write().unwrap();
-                    match &*guard {
-                        Some(current_field_value) => {
-                            if Arc::ptr_eq(current_field_value, &proc) {
+                    match self.redo_process.get() {
+                        None => (),
```

Comment:
I would prefer to keep the enum order for these match arms.

I was also wondering is the None branch here ok.

I think it is correct:

1. first one observing an error will get true for `Arc::ptr_eq`, take the permit and hold it while doing `kill; wait;`
2. while the permit is being held, `get` will return `None`, so the rest of "error observers" will continue fast from here
3. until the deinit is ready, all retryers will wait to get the semaphore

Yes I think it works as it should here.

Reply:
Actually, there might be an implicit drop-order thing going on.

`Guard::take_and_deinit` returns `(T, InitPermit)`. I think that value is dropped like:

1. `T`
2. `InitPermit`

However, we could make this explicit by:

```patch
-guard.take_and_deinit();
+let (proc, permit) = guard.take_and_deinit();
+drop(proc);
+drop(permit);
```

This is exactly what the implicit version is already doing per my understanding ([reference](https://doc.rust-lang.org/reference/destructors.html)):
> The fields of a [tuple](https://doc.rust-lang.org/reference/types/tuple.html) are dropped in order.

But here it really matters to us that we hold the permit over the `wait+kill`, correct?

Reply:
Aah I missed there was the line L303R306 with `drop(proc)`.

Ok, so I had misunderstood: because `proc` is an Arc, we cannot guarantee that a drop is happening while we are holding the init permit, and perhaps that was not the aim here at all, I was just imaging it was?

Reply:
> Ok, so I had misunderstood: because proc is an Arc, we cannot guarantee that a drop is happening while we are holding the init permit,

Exactly, this should be evident from the block comment starting in line 294 as well as the comment on field `redo_process`; but apparently it's only evident for me.

I'll try to rearrange the code a bit to make it more clear

Reply:
ok yes I was partially remembering correctly. I didn't remember we wanted to accept N=async threads walredo processes.

well, it doesn't really matter, occurrences are so rare.